### PR TITLE
Fixed #32813 -- Display development server address after server bind

### PR DESCRIPTION
**Pull request by TheAbhijeet**
## Changelog

So far Django displays the server name and port before actual binding;

If a user selects port 0, [wildcard port]:
```
$ ./manage.py runserver 127.0.0.1:0
...
Starting development server at http://127.0.0.1:0/
Quit the server with CONTROL-C.
```
Now if the user opens a browser with this address they will not find the app running here. Because for the wildcard port, the actual physical port is selected by the operating system.  

After this change Django will be able to pick up the correct port:
```
$ ./manage.py runserver 127.0.0.1:0
...
Starting development server at http://127.0.0.1:25837/
Quit the server with CONTROL-C.
```

Ref - https://code.djangoproject.com/ticket/32813

### Backstory

I was looking for tickets for my first contribution to Django and I landed on this task, the original PR https://github.com/django/django/pull/14250 was closed so I picked it up from there.

### CODE REVIEW

File: tests/admin_scripts/tests.py [hunk 1]
```diff
@@ -2363,7 +2373,6 @@ def test_program_name_from_argv(self):
 
 @override_settings(ROOT_URLCONF="admin_scripts.urls")
 class StartProject(LiveServerTestCase, AdminScriptTestCase):
-
```
Comment (carltongibson) [inline tests/admin_scripts/tests.py:2366]: Please remove this (and other) unrelated change(s).

**Review by carltongibson — COMMENTED** [review]
Hi @TheAbhijeet — Welcome aboard! ⛵ :)

Can I ask you to look at the reviews from #14250? In particular Nick suggested f-strings for the string formatting there, which is probably a good move. 

Then I think it's probably not far off.

**Comment by carltongibson** [comment]
I've checked _Patch needs improvement_ on the ticket-32813. Please uncheck that when you've updated 👍

File: tests/admin_scripts/tests.py (hunk 1 above)
Comment (TheAbhijeet) [reply tests/admin_scripts/tests.py]: Thanks for the feedback, I have made the changes.

**Comment by TheAbhijeet** [comment]
Hi @carltongibson @felixxm 

I don't understand why MySQL builds are failing. Can you please help me solve this?

**Comment by TheAbhijeet** [comment]
> Hi @carltongibson @felixxm
> 
> I don't understand why MySQL builds are failing. Can you please help me solve this?

Apparently re-running the test cases worked.

File: django/core/servers/basehttp.py [hunk 2]
```diff
@@ -227,13 +227,30 @@ def handle_one_request(self):
...
-        httpd_cls = server_cls
+    cls_dict = {}
+
+    if on_bind is not None:
+
+        def server_bind(self):
+            server_cls.server_bind(self)
+            on_bind(self.server_port)
+
+    cls_dict["server_bind"] = server_bind
```
Comment (carltongibson) [inline django/core/servers/basehttp.py:248]: It looks to me as if `server_bind` would be referenced before  assignment when `on_bind is None`. 

Can you please add a test case for that scenario, or else we'd need to make the parameter required if it were always passed. 🤔

File: AUTHORS [hunk 3]
```diff
@@ -9,6 +9,7 @@ answer newbie questions, and generally made Django that much better:
     Aaron Swartz <http://www.aaronsw.com/>
     Aaron T. Myers <atmyers@gmail.com>
     Abeer Upadhyay <ab.esquarer@gmail.com>
+    Abhijeet Pal
```
Comment (carltongibson) [inline AUTHORS:12]: Do you want to add your email?

**Review by carltongibson — COMMENTED** [review]
Just one query about the `server_bind` block... 

Can you add: a `Co-authored-by: Ali Farzanrad <ali_farzanrad@riseup.net>` to the commit message please. 

Thanks!

File: AUTHORS (hunk 3 above)
Comment (TheAbhijeet) [reply AUTHORS:12]: Naa, it's fine without it.

File: django/core/servers/basehttp.py (hunk 2 above)
Comment (TheAbhijeet) [reply django/core/servers/basehttp.py:248]: Apparently, on_bind is always passed, so I made it a required parameter.

**Comment by TheAbhijeet** [comment]
Updated the branch, please let me know what you think.

File: django/core/servers/basehttp.py (hunk 2 above)
Comment (TheAbhijeet) [reply django/core/servers/basehttp.py]: Hi, @carltongibson sorry for the confusion for some reason my changes where I made it a required weren't pushed. 

However, I have made `on_bind` a required argument now.

**Comment by felixxm** [comment]
Updated in #16319.
